src/internal/parallel/each_test.go

```diff
@@ -0,0 +1,58 @@
+package parallel
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestForEach(t *testing.T) {
+	// Note: these actions are not thread-safe, but that's fine for this test
+	// The idea of this test is to ensure that the concurrency is working as expected
+	// and that the errors are returned as expected
+
+	successAction := func() error {
+		time.Sleep(50 * time.Millisecond) // simulate some work
+		return nil
+	}
+
+	errorAction := func() error {
+		time.Sleep(50 * time.Millisecond) // simulate some work
+		return errors.New("error occurred")
+	}
+
+	tests := []struct {
+		name           string
+		actions        []Action
+		maxConcurrency int
+		expectedErrs   int
+	}{
+		{
+			name:           "All successful actions",
+			actions:        []Action{successAction, successAction, successAction},
+			maxConcurrency: 2,
+			expectedErrs:   0,
+		},
+		{
+			name:           "Some actions with errors",
+			actions:        []Action{errorAction, successAction, errorAction},
+			maxConcurrency: 2,
+			expectedErrs:   2,
+		},
+		{
+			name:           "Some actions with errors, more actions than concurrency",
+			actions:        []Action{errorAction, successAction, errorAction, errorAction, errorAction, errorAction},
+			maxConcurrency: 2,
+			expectedErrs:   5,
+		},
```

Comment:
It might make more sense to make this a generic function, similar to @cube2222 's suggestion in the original PR that added this.  I didn't want to add that at the time, since it would have been the first generics added to the build/codebase.

Reply:
The java equiv might look like: `public static <T> List<Pair<T, Error>> ForEach(List<T> items, Function<T, Error> action)`.  I need to spend some time learning about generics in go

Reply:
It could also be done with any/casting as this is not currently used by a tight loop

Reply:
https://github.com/opentofu/registry-stable/issues/40